exercises/practice/linked-list/.meta/example.ex

```diff
@@ -0,0 +1,53 @@
+defmodule LinkedList do
+  defstruct front: [], back: []
+
+  @typedoc """
+  A linked list is made of two singly linked lists at the front and back
+  """
+  @type linked_list :: %LinkedList{front: [any], back: [any]}
+
+  @doc """
+  Create a new linked list
+  """
+  @spec new() :: linked_list
+  def new, do: %LinkedList{}
+
+  @doc """
+  Insert a value at the back
+  """
+  @spec push(linked :: linked_list, value :: any) :: linked_list
+  def push(%LinkedList{back: back} = linked, value), do: %{linked | back: [value | back]}
+
+  @doc """
+  Remove a value at the back
+  """
+  @spec pop(linked :: linked_list) :: {any, linked_list}
+  def pop(%LinkedList{back: [value | back]} = linked), do: {value, %{linked | back: back}}
+  def pop(%LinkedList{front: front}), do: pop(%LinkedList{back: Enum.reverse(front)})
```

Comment:
Reversing a part of the linked list when removing its first element somehow feels wrong. Like... against the whole concept of linked lists? I'm not an expert on this, I quit my CS course after a year 😊 

I need to think about this a bit longer. Double linked lists and Elixir, hmm. Could they be faked better with structs and maps? 

E.g. a linked list would be a struct with three fields: start, end, and nodes:
```
%LinkedList{
  start: 4,
  end: 19,
  nodes: %{
    4: %{value: "banana", next: 7, prev: 19},
    7: %{value: "apple", next: 19, prev: 4},
    19: %{value: "orange", next: 7, prev: 7},
  }
}
```

The numbers are some random identifiers.

Reply:
Yea, that would probably more akin to how I might solve it but @jiegillet's solution still works, so still good by me

Reply:
Well, of course you can use a map to simulate a linked list, or you could even more simply use a tuple:
```
linked = {"banana", "apple", "orange"}
```
and use `Tuple.insert_at`, `Tuple.delete_at` for `push` and `pop`, but using a very sophisticated built-in data structure defeats the purpose of implementing your own data structure. 

Using a singly linked list to implement a doubly linked list (or really, a queue) is a more fair game. Yes, indeed, I need to reverse lists once in a while, which is not in the spirit of singly linked lists (although I think if you study the complexity, it's still not so bad because you don't need to do that very often), but it's [standard](http://erlang.org/doc/man/queue.html).

It feels like students might be confused by that exercise, because many of them didn't study CS and even less studied functional data structures. So I'm thinking either we drop it, because doubly linked lists as they are defined are impossible in immutable languages, or we give more instructions on what we expect the student to do (after deciding what that is).

Reply:
> because doubly linked lists as they are defined are impossible in immutable languages

In that case I would say we should drop it. We have enough nice exercises, and I don't want to encourage people to do things that don't make much sense in Elixir.

We already had a few complaints about `binary-search` because the only way to implement it while keeping its intended complexity is to use a tuple to store a collection, but you wouldn't do that in Elixir. Tuples are not for collections, but you can't do binary search on a list.

Reply:
OK, sounds good to me. 
If there ever is an exercise to "build a queue" then we dig for this implementation.

Reply:
To mark an exercise as "please do not implement now or ever", we would need to add it to the `"foregone"` list in `config.json`. There's already one exercise there.